app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/DeviceListFragment.java

```diff
@@ -0,0 +1,357 @@
+
+  public static final String TAG = "DeviceListFragment";
+  public static int PEER_HANDSHAKE_PORT = 8009;
+  public static int FILE_TRANSFER_PORT = 8008;
+
+  private SharedPreferenceUtil sharedPreferenceUtil;
+
+  private LocalFileTransferActivity localFileTransferActivity;  // Parent activity, starting point of the module
+  private TransferProgressFragment transferProgressFragment;    // Sibling fragment, for displaying transfer progress
+
+  private View deviceListFragmentRootView = null; // Root view corresponding to the DeviceListFragment
+
+  private boolean fileSender = false; // Whether file sending device or not
+  private ArrayList<Uri> fileUriList; // Uris of files to be shared, available only for the sender device
+  private int totalFilesForTransfer = -1;
+  private int filesSent = 0;          // Count of number of files transferred until now
+  private ArrayList<FileItem> filesToSend = new ArrayList<>();
+
+  private WifiP2pDevice userDevice;   // Represents the device on which the app is running
+  private WifiP2pInfo groupInfo;      // Corresponds to the WiFi P2P group formed between the two devices
+  private List<WifiP2pDevice> peerDevices = new ArrayList<WifiP2pDevice>();
+
+  private WifiP2pDevice selectedPeerDevice = null;
+  private InetAddress selectedPeerDeviceInetAddress;
+
+  private InetAddress fileReceiverDeviceAddress;  // IP address of the file receiving device
+  private boolean fileTransferStarted = false;
+
+  @Override
+  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+    super.onActivityCreated(savedInstanceState);
+    localFileTransferActivity = (LocalFileTransferActivity) getActivity();
+
+    // As DeviceListFragment extends ListFragment for the purpose of displaying list of peers
+    this.setListAdapter(new WifiPeerListAdapter(localFileTransferActivity, R.layout.row_peer_device, peerDevices));
+
+    if(localFileTransferActivity.isFileSender()) {
+      fileSender = true;
+      fileUriList = localFileTransferActivity.getFileUriArrayList();
+      totalFilesForTransfer = fileUriList.size();
+
+      for(int i = 0; i < fileUriList.size(); i++) {
+        filesToSend.add(new FileItem(getFileName(fileUriList.get(i)), FileItem.TO_BE_SENT));
+      }
+
+      displayTransferProgressFragment();
+    }
+
+  }
+
+  @Override
+  public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+    deviceListFragmentRootView = inflater.inflate(R.layout.fragment_device_list, null);
+    return deviceListFragmentRootView;
+  }
+
+  @Override
+  public void onListItemClick(ListView l, View v, int position, long id) {
+    /* Connection can only be initiated by user of the sender device, & only when transfer has not been started */
+    if(!isFileSender() || fileTransferStarted)
+      return;
+
+    selectedPeerDevice = (WifiP2pDevice) getListAdapter().getItem(position);
+    new AlertDialog.Builder(localFileTransferActivity)
```

Comment:
Please see the class AlertDialogShower, I would like for all alertDialogs to use this

Reply:
@macgills I'm not clear on how to use AlertDialogShower
Like I have added the following to KiwixDialog.kt:
```
object LocationPermission : KiwixDialog(
      null, R.string.permission_rationale_location, R.string.yes, R.string.no
  )
```
Now to show the dialog, I'm writing:
```
@Inject DialogShower dialogShower;
.
.
.
dialogShower = new AlertDialogShower(this, sharedPreferenceUtil);
dialogShower.show(new KiwixDialog.LocationPermission(), requestPermissions(new String[]{Manifest.permission.ACCESS_COARSE_LOCATION}, PERMISSION_REQUEST_CODE_COARSE_LOCATION));
```
where `new KiwixDialog.LocationPermission()` is underlined with a "_'LocationPermission()' has private access in 'org.kiwix.kiwixmobile.utils.KiwixDialog.LocationPermission'_" error
I'm probably using the injection wrong, but I can't understand how the kotlin code is working

How should I go about this?

Reply:
Do not assign to an injected field. When you call `inject` the fields that have that annotation will receive an instance of the correct class.
I did a quick java example
```
   dialogShower.show(KiwixDialog.DeleteZim.INSTANCE, new Function0<Unit>() {
          @Override public Unit invoke() {
            //do the positive action
            return Unit.INSTANCE;
          }
        },
        new Function0<Unit>() {
          @Override public Unit invoke() {
            //do the negative action
            return Unit.INSTANCE;
          }
        });
```
it is much much uglier in java unfortunately

Reply:
> Do not assign to an injected field.

Don't know what was I thinking doing that

Thanks for the example :+1:

Reply:
Is it possible to use template strings with `AlertDialogShower`? Otherwise I'll stick to AlertDialogBuilder for the particular case

Reply:
I now have an example for doing this with DeleteZim. I did it as an interface and one of the rather cool things in kotlin is you can override an abstract property and implement it as a constructor parameter

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/544fcd2215f282e15c38b3c6aa2ace928da74949)
Not sure whether I followed the DeleteZim example correctly, could you have a look [here](https://github.com/kiwix/kiwix-android/pull/1276/commits/544fcd2215f282e15c38b3c6aa2ace928da74949#diff-81eda37ace31e1bdb5df37c7fb95dcb9R28)?